### PR TITLE
#40 Too many open files

### PR DESCRIPTION
**Pull request by magdel**
Via AutoCloseable

### CODE REVIEW

**Comment by 0crat** [comment]
@yegor256 please, pay attention to this pull request

**Comment by 0crat** [comment]
Job `gh:yegor256/jpeek#51` is in scope.

File: src/test/resources/org/jpeek/metrics/SkippedEnum.java [hunk 1]
```diff
@@ -0,0 +1,4 @@
+public enum SkippedEnum {
```
Comment (alex-semenyuk) [inline src/test/resources/org/jpeek/metrics/SkippedEnum.java]: Why this is needed?

File: src/test/resources/org/jpeek/metrics/SkippedAnnotation.java [hunk 2]
```diff
@@ -0,0 +1,2 @@
+public @interface SkippedAnnotation {
+}
```
Comment (alex-semenyuk) [inline src/test/resources/org/jpeek/metrics/SkippedAnnotation.java]: Why this is needed?

File: src/test/resources/org/jpeek/metrics/SkippedEnum.java (hunk 1 above)
Comment (magdel) [reply src/test/resources/org/jpeek/metrics/SkippedEnum.java]: Codecov shows not covered lines
 && !ctClass.isEnum()
&& !ctClass.isAnnotation()
so it was a try to grow up coverage.
But I don't see any enhancement there(

Comment (yegor256) [reply src/test/resources/org/jpeek/metrics/SkippedEnum.java]: @magdel this looks weird :) please, remove these files

**Comment by baumato** [comment]
Hi.
I just had a look into this issue, because I wanted to fix it and found this already created patch. The solution provided here seems to be correct (not taking SkippedEnum and SkippedAnnotation into account).
The problem is, that a try-with-resource statement cannot be tested easily (or at all?) because it gets compiled to a lot of branches. For further information see here: https://github.com/jacoco/jacoco/issues/82

So the options are to suppress the issue or not using try-with-resources at all (e.g. using guava's Closer which avoids to introduce a null initialized variable).

Best regards,
Tobias

File: src/test/resources/org/jpeek/metrics/SkippedEnum.java (hunk 1 above)
Comment (magdel) [reply src/test/resources/org/jpeek/metrics/SkippedEnum.java]: ok

**Comment by alex-semenyuk** [comment]
Is it so important have 100% code coverage instead of using Guava static method?
Totally agree with this 
"While 100% code coverage is nice, it's not a good metric to blindly follow. 100% code coverage doesn't tell you anything other than you have tests that execute all your lines of code. Your code can still have bugs if that covered code does the wrong thing or doesn't do enough."(c ) https://stackoverflow.com/questions/26360245/try-with-resource-unit-test-coverage

**Comment by magdel** [comment]
Both code versions with try-with-resorce and with Closer are valid but fail with coverage or codestyle.
Left version with successful build.

**Comment by yegor256** [comment]
@rultor try to merge

**Comment by magdel** [comment]
Fixed checkstyle issues.

**Comment by magdel** [comment]
Well, is there a way to merge the fix?
Using try-with-resource we a doomed to decrease coverage (would be crazy idea to cover all generated branches because we will need to inject factory of fake FileInputStream creating all necessary behaviors)

**Comment by yegor256** [comment]
@rultor merge

**Comment by yegor256** [comment]
@magdel every time you say something in GitHub, start your message with the name of the person you're addressing it to. Otherwise, they simply won't see your messages.

**Comment by yegor256** [comment]
@magdel thanks!

**Comment by magdel** [comment]
@yegor256 @alex-semenyuk @baumato thank you all, I've learned much

**Comment by yegor256** [comment]
@magdel you may find this useful too: http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html

**Pull request merged**